### PR TITLE
Added Catalyst2 Build Support

### PR DESCRIPTION
**Pull request by andrewcombs**
## Summary
This is an equivalent implementation of Catalyst 2 to that of Ascent. It adds it to the build system so that it can be exposed to downstream codebases.

## Additional background
Catalyst 2 is adding support for AMR-based datasets. Part of that initiative is full integration of the Catalyst 2 in situ analysis and visualization platform into AMReX. A [pull request to WarpX](https://github.com/ECP-WarpX/WarpX/pull/4121) is live demonstrating that implementation.

## Checklists

TODO:
- [ ] Add a github workflow test

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [X] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

**Comment by c-wetterer-nelson** [comment]
Thank you @andrewcombs!

@ax3l Andrew is an intern I have been advising here at Kitware. He has been implementing Catalyst 2 functionality here in AMReX and downstream in WarpX. We have a few to dos left before it will be ready to merge. Is there anything missing we should add to the to do list? Thanks!

**Comment by ax3l** [comment]
This is great, thank you for the heads-up and great work!

As one detail, do you think you can add the generalized particle container support as in #3349 already? (Same need to go into #3350 still).

We are about to switch to that data format in WarpX and in situ vis are the last components that need migrating:
https://github.com/ECP-WarpX/WarpX/pull/3850

File: Src/Extern/Conduit/CMakeLists.txt [hunk 1]
```diff
@@ -4,6 +4,7 @@ foreach(D IN LISTS AMReX_SPACEDIM)
     # TODO: Particles PR merges another file
     target_sources(amrex_${D}d
        PRIVATE
+       AMReX_Conduit_Blueprint_ParticlesI.H
```
Comment (WeiqunZhang) [inline Src/Extern/Conduit/CMakeLists.txt:7]: Similarly to this, could you also update `Make.package` for GNU Make?

**Comment by WeiqunZhang** [comment]
@andrewcombs For the TODOs, do you plan to do it in this PR or another PR?

**Comment by andrewcombs** [comment]
Likely in this PR, within the next week if time allows

**Comment by c-wetterer-nelson** [comment]
~~whoops. looks like I built the catalyst image for arm64. fixing that now.~~
updated the container to amd64. hopefully that'll let the actions run. gotta rerun the new catalyst build test now.

**Comment by ax3l** [comment]
@c-wetterer-nelson @andrewcombs quick question: this PR seems to only link against Catalyst, but add no helpers to AMReX.

We can certainly merge it, but I am surprised there are not helpers (e.g., as in Ascent) being added here. Otherwise the pure find package logic for Catalyst can also be done downstream (in WarpX), no?

**Review by ax3l — APPROVED** [review]
Oh, I see now. You essentially pick up the logic from Conduit? Good for me.

**Comment by ax3l** [comment]
Cycling CI one last time to make sure it still passes.

File: .github/workflows/catalyst.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    name: Catalyst
+    runs-on: ubuntu-22.04
+    if: github.event.pull_request.draft == false
+    env:
+      CXX: g++
+      CC: gcc
+    container:
+      image: cwetterernelson/catalyst:v2
+    steps:
+    - uses: actions/checkout@v3
```
Comment (ax3l) [inline .github/workflows/catalyst.yml]:
```suggestion
    - uses: actions/checkout@v4
```

File: .github/workflows/catalyst.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    - name: Configure
+      run: |
+        cmake -S . -B build          \
```
Comment (ax3l) [inline .github/workflows/catalyst.yml:25]: @c-wetterer-nelson I think your docker container might have a small bug. It sets an env variable `catalyst_DIR`, but I think it needs to be case sensitive:
```suggestion
        export Catalyst_DIR=${catalyst_DIR}
        cmake -S . -B build          \
```
https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure

```
CMake Error at Tools/CMake/AMReXThirdPartyLibraries.cmake:97 (find_package):
  Could not find a package configuration file provided by "Catalyst" with any
  of the following names:

    CatalystConfig.cmake
    catalyst-config.cmake
```

File: .github/workflows/catalyst.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      CXX: g++
+      CC: gcc
+    container:
+      image: cwetterernelson/catalyst:v2
+    steps:
+    - uses: actions/checkout@v4
+    - name: Configure
+      run: |
+        export Catalyst_DIR=${catalyst_DIR}
+        export Catalyst_DIR=${catalyst_DIR}
```
Comment (ax3l) [inline .github/workflows/catalyst.yml]:
```suggestion
        export Conduit_DIR=${Catalyst_DIR}
```

File: .github/workflows/catalyst.yml (hunk 3 above)
Comment (ax3l) [reply .github/workflows/catalyst.yml:25]: An env hint for `Conduit_DIR` also needs to be set :)

**Comment by c-wetterer-nelson** [comment]
Hi @ax3l I'm just getting to this now. Let me take a look at the docker image and make sure it's all working. You're right that this does not add a lot of content outside of the compile logic, and picks up the Conduit logic thanks to Catalyst's closeness to Ascent. 

We also have the PR in WarpX that relies on this PR. https://github.com/ECP-WarpX/WarpX/pull/4121/files

**Comment by ax3l** [comment]
Thanks @c-wetterer-nelson,

Yes, let's ensure this PR compiles and then we take on the WarpX one :) Can you see the CI output here and pls update the docker container as needed? :)

**Comment by ChristosT** [comment]
Hey @ax3l I will be taking this over. I believe I cannot push to to this PR so I created a new one #4011 . Let me know if you have any feedback.Thanks !

**Pull request closed without merging**